### DISCUSSION TITLE
Versioning aggregate roots based on date time

### DISCUSSION THREAD

**Discussion by stevewolfe555** [spatie/laravel-event-sourcing#483] [Q&A]
I'm working on their system that receives events delayed and out of time but they all come with a timestamp.

I need (I think) the aggregate root to be rebuilt based on this timestamp so I have extended the Should be stored and the stored event repository.

This however leaves me an issue that when you try to rebuild an aggregate root it's still trying to get the version.

I was trying to remove the version IDs as these will no longer be accurate.

I was looking to replace the aggregate root class but thought I would check to see if anyone has had an experience with the best way to deal with this.

My business logic will have to understand that if an event comes in out of order it will have to rebuild projections but I'm still trying to keep the stored events table orderable so that when it's retrieved on the aggregate root everything is correct.

**Comment by inmanturbo** [comment]
Rebuilding projections is one thing, but reordering the events is another entirely.

How out of order? Is there a threshold? Or if an event comes in with a timestamp from two years ago would you rebuild projections and try to put that event ahead of all the others in the last two years?

I ask because if there is a threshold, and it's much shorter than two years (as I would expect), say 10 minutes? Then I would cache the events with a ten minute lifespan (from the time of their timestamp) and flush them to the store in the correct order every say ... 10 minutes.

**Reply by stevewolfe555** [reply]
So i think there are instances where they could be weeks out of date.

What i have done is added an event date column in addition to the created column. Then I can retrieve events on the aggregate route by the event date. 

Then what I'm doing when it comes to my projections and reactors is comparing if the event date is before the last updated day on that projection. If it is then I rebuild the projection from the aggregate route if it needs to be other wise let it carry on.

Only some events need to rebuild a projections others congest insert them in for example a history table.

**Reply by Gummibeer** [reply]
#515 wasn't intended for that but could help you with that. Just use your timestamp column as the order by and all queries will automatically retrieve and apply the events in the timestamp order.
